macrosynergy/management/dq.py

```diff
@@ -419,6 +433,58 @@ def _fetch_ts(self, endpoint: str, params: dict,
 
         return results
 
+    def _optimize(self, endpoint: str, tickers: List[str], params: dict,
+                  start_date: str = None, end_date: str = None,
+                  calendar: str = "CAL_ALLDAYS", frequency: str = "FREQ_DAY",
+                  conversion: str = "CONV_LASTBUS_ABS",
+                  nan_treatment: str = "NA_NOTHING"):
+
+        params_ = {"format": "JSON", "start-date": start_date, "end-date": end_date,
+                   "calendar": calendar, "frequency": frequency, "conversion":
+                   conversion, "nan_treatment": nan_treatment}
+        params.update(params_)
+
+        no_tickers = len(tickers)
+        iterations = ceil(no_tickers / 20)
+        remainder = no_tickers % 20
+
+        tick_list_compr = []
+        for i in range(iterations):
+            if i < (iterations - 1):
+                tick_list_compr.append(tickers[i * 20: (i * 20) + 20])
+            else:
+                tick_list_compr.append(tickers[-remainder:])
+
+        final_output = []
+        output = []
+        if self.concurrent:
+            with concurrent.futures.ThreadPoolExecutor() as executor:
+                for elem in tick_list_compr:
+                    params["expressions"] = elem
+                    results = executor.submit(self._fetch_threading, endpoint, params)
+                    time.sleep(0.75)
+                    output.append(results)
+
+                for f in concurrent.futures.as_completed(output):
```

Comment:
Yep, otherwise the threads are engaged serially: according to the order of the iterator.

The second Loop allows for iterating through the threads as they're returned. For instance:

Current Thread: <Thread(ThreadPoolExecutor-0_6, started daemon 123145616498688)>.
Current Thread: <Thread(ThreadPoolExecutor-0_5, started daemon 123145599709184)>.
Current Thread: <Thread(ThreadPoolExecutor-0_1, started daemon 123145532551168)>.
Current Thread: <Thread(ThreadPoolExecutor-0_0, started daemon 123145514688512)>.
Current Thread: <Thread(ThreadPoolExecutor-0_2, started daemon 123145549340672)>.
Current Thread: <Thread(ThreadPoolExecutor-0_3, started daemon 123145566130176)>.
Current Thread: <Thread(ThreadPoolExecutor-0_9, started daemon 123145666867200)>.
Current Thread: <Thread(ThreadPoolExecutor-0_8, started daemon 123145650077696)>.
Current Thread: <Thread(ThreadPoolExecutor-0_4, started daemon 123145582919680)>.
Current Thread: <Thread(ThreadPoolExecutor-0_7, started daemon 123145633288192)>.

If contained in a single Loop, it would defeat the purpose of threading.